src/d3a/models/area/markets.py

```diff
@@ -15,84 +15,158 @@
+            self._move_settlement_markets_to_past_settlement(current_time)
+            self._delete_past_settlement_markets(current_time)
+
+        self._update_indexed_future_markets()
+
+    def _move_markets_to_past(self, current_time: DateTime) -> None:
+        """Move slot markets to self.past_markets."""
+        # the conversion to list is needed in order to omit 'mutated during iteration' error
+        for time_slot in list(self.markets.keys()):
+            if time_slot < current_time:
+                market = self.markets.pop(time_slot)
                 market.readonly = True
-                self._delete_past_markets(past_markets)
-                past_markets[timeframe] = market
-                self.log.trace("Moving {t:%H:%M} {m} to past"
-                               .format(t=timeframe, m=past_markets[timeframe].name))
-
-    def _delete_past_markets(self, past_markets):
-        if not d3a.constants.RETAIN_PAST_MARKET_STRATEGIES_STATE:
-            delete_markets = [pm for pm in past_markets if
-                              pm not in self.markets.values()]
-            for pm in delete_markets:
-                if ConstSettings.GeneralSettings.EVENT_DISPATCHING_VIA_REDIS:
-                    past_markets[pm].redis_api.stop()
-                del past_markets[pm].offers
-                del past_markets[pm].trades
-                del past_markets[pm].offer_history
-                del past_markets[pm].notification_listeners
-                del past_markets[pm].bids
-                del past_markets[pm].bid_history
-                del past_markets[pm].traded_energy
-                del past_markets[pm]
+                self.past_markets[time_slot] = market
+                self.log.debug(
+                    f"Moving {self.past_markets[time_slot]} to past.")
+
+    def _move_balancing_markets_to_past_balancing(self, current_time: DateTime) -> None:
+        """Move balancing markets to self.past_balancing_markets."""
+        # the conversion to list is needed in order to omit 'mutated during iteration' error
+        for time_slot in list(self.balancing_markets.keys()):
+            if time_slot < current_time:
+                market = self.balancing_markets.pop(time_slot)
+                market.readonly = True
+                self.past_balancing_markets[time_slot] = market
+                self.log.debug(
+                    f"Moving {self.past_balancing_markets[time_slot]} to past.")
+
+    def _move_settlement_markets_to_past_settlement(self, current_time: DateTime) -> None:
+        """Move settlement markets to self.past_settlement_markets."""
+        move_settlement_market_slots = [
+            time_slot for time_slot in self.settlement_markets.keys()
+            if self._is_it_time_to_delete_past_market(current_time, time_slot)]
+        for time_slot in move_settlement_market_slots:
+            self.past_settlement_markets[time_slot] = self.settlement_markets.pop(time_slot)
```

Comment:
we should only have one editable settlement market (the latest), and the rest should be readonly, right ?

Reply:
No, the participants should be able to place bids and offers in all of the settlement markets as along as they are not in `past_settlement_markets.`